S49P15 #11 Upgrade Admin Server: UI Story 2 - Service Integration

Itxaka opened the pull request:
- map the response data to the model properly
- launch the check as soon as the model is created, to check for ongoing updates
- Add proper responses to specs
- Change the button text on upgrade complete
- Extra check before setting the timeout

Commit:
- map the response data to the model properly
- launch the check as soon as the model is created, to check for ongoing updates
- Add proper responses to specs
- Change the button text on upgrade complete
- Extra check before setting the timeout

Itxaka commented:
Criteria was:
A mechanism to start the admin server upgrade, and to check when it is finished is implemented.
Expected temporary API downtime (e.g. when the admin server is rebooting) is handled properly (i.e. the error seen by the UI during this time window is not treated as an ordinary error)

In my tests I have found:
- Start the upgrade -> Already done and in place
- Check when its finished -> was almost done and only required a couple of canges to the reponse workflow
- Temporary downtime -> When the node is down the service already keeps checking. There is nothing in the UI to indicate that there is an error anywhere.

Problems with this:
- If you launch the upgrade and refresh the page, the state is lost. I added a first check on model initialisation so if the upgrade is ongoing it will disable the button and start the periodic check for status.
-  Im still not sure what kind of errors the API can return nor if it can even return anything (returns 422 if you try to start the upgrade and the upgrade is ongoing but that cant happen anymore)
- Next button is enabled all the time, nor there is anything in the model to flag the current step as finished so the button is enabled on step completition.

Itxaka commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) S49P15 #11 Upgrade Admin Server: UI Story 2 - Service Integration](https://trello.com/c/PAkqYMEe/147-3-s49p15-11-upgrade-admin-server-ui-story-2-service-integration)

Itxaka commented:
Oh yeah, on testing, due to the controller being created beforeEach at the top, I am unable to override the `initialResponse` called on controller creation (the first call to `checkAdminUpgrade()`).

I would like to be able to modify this for a subset of tests in order to initialise the model with different initial statuses. Other than creating the controller per `describe` section, Im not sure how to do it. Any thoughts?

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.js:
```diff
@@ -24,6 +24,10 @@
 
+        // On page load we need to check to see if the upgrade is already running to set
+        // the button status and the update check running
+        checkAdminUpgrade();
```
Could you please use an activate() function here?
https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#controller-activation-promises
Also, to keep consistency with the Nodes Repo checks controller:
https://github.com/crowbar/crowbar-ui/pull/58/files#diff-2230c16354311fde135c40b1f3902a44R72

Itxaka replied:
Will do, thanks!

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.js:
```diff
@@ -48,21 +52,25 @@
+                        } else if (vm.adminUpgrade.completed && !vm.adminUpgrade.running) {
+                            // TODO(itxaka): can translations be done from here instead so
+                            // we can update the button text for better UX?
+                            vm.finish_text = 'Upgrade done!'
```
The translation service can be injected into controllers and used to send text to the view (see https://angular-translate.github.io/docs/#/api/pascalprecht.translate.$translate), but I'd rather use such case only if no other option is possible.
I'd prefer using the VM to expose the state to the view and let it decide what to show, where and when.

Another possibility is to let the controller build and update language keys to be exposed to the view, and use the translations there.

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.js:
```diff
@@ -48,21 +52,25 @@
                     function () {
                         // schedule another check if not completed yet
-                        if (!vm.adminUpgrade.completed) {
+                        if (!vm.adminUpgrade.completed && vm.adminUpgrade.running) {
```
Now that I see both in a single condition, doesn't each of those attributes suggest the opposite value in the other already? Sounds a bit redundant to me.
What do you think @skazi0 ?

Itxaka replied:
This is done to avoid the timeout from firing on page load, as we are now calling the checkAdminUpgrade on controller creation, it wiould always trigger it. Instead we check both values so it only triggers it when the upgrade is ongoing.

skazi0 replied:
They are not direct opposite of each other. `completed` means a complete run was finished (success or fail). `running` means something is currently running. Before anything is started it's neither `completed` nor `running`.

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.spec.js:
```diff
@@ -2,12 +2,36 @@
+                upgrade: {
+                    upgrading: false,
+                    success: true,
+                    failed: false}
```
Please, closing brace in a new line

Itxaka replied:
will do

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.spec.js:
```diff
@@ -2,12 +2,36 @@
+                upgrade: {
+                    upgrading: true,
+                    success: false,
+                    failed: false}
```
same here

Itxaka replied:
will do

santiph commented on assets/app/features/upgrade/templates/upgrade-administration-server-page.jade:
```diff
@@ -10,4 +10,4 @@ button.btn.btn-success.center-block(
     suse-lazy-spinner(delay="2000", active="upgradeAdminUpgradeVm.adminUpgrade.running",
         visible="upgradeAdminUpgradeVm.adminUpgrade.spinnerVisible")
-    span(translate='') upgrade.steps.upgrade-admin.form.upgrade-admin
+    span(translate='', ng-bind="upgradeAdminUpgradeVm.finish_text") upgrade.steps.upgrade-admin.form.upgrade-admin
```
Doesn't ngBind generate an unnecessary watcher?

Itxaka replied:
This was only done as a test for the text change on upgrade finish and for the translation question. It will be removed in the next comimt :)

santiph commented:
@Itxaka Don't worry on the errors the API could eventually return and/or the case the API doesn't come back at all. It's not part of the scope of this card and we have a specific one for Error Handling for this page: https://trello.com/c/kxvh6yP2/18-upgrade-admin-server-error-handling-story-1 (Still to be refined, tho).
Same thing with Next button being enabled/disabled. That should be cover as part of the navigation bar user story: https://trello.com/c/LNr710jO/104-3-s49p20-6-layout-navigation-bar

There is a static service in place returning fake data for the UI that you could use to simulate the downtime on the Admin node and see how angular would respond to that: https://github.com/crowbar/crowbar-ui/blob/master/routes/api/crowbar/upgrade.js
Would you mind to update the static responses also to match the documentation?

Finally, I left a couple of comments on the review... Looks good so far

Commit:
- Update the fake data upgrade endpoint to return the proper format and simulate transitioning statuses
 - Fix style issues
 - Use activate function for clarity

santiph commented on assets/app/features/upgrade/controllers/upgrade-upgrade-administration-server.controller.js:
```diff
@@ -24,6 +24,14 @@
+        function activate() {
+            // On page load we need to check to see if the upgrade is already running to set
+            // the button status and the update check running
+            checkAdminUpgrade();
```
We need to test this functionality, somehow. I'm still not sure about the proper approach for this that we should follow here (and in the following steps). But is not part of the scope of the current card we've assigned this sprint.
It's part of Restore last step one, as detailed in https://trello.com/c/5fXGm1a7/45-2-27-restore-last-step (Has not been properly refined, estimated, prioritised nor committed yet. ;) )

santiph replied:
Don't remove it, just add a comment saying it will be properly implemented and tested as part of that card

santiph commented:
Looks good to me. Please address the comment regarding checking the current status and consider this approved

Commit:
- Update error TODO to clarify the reasoning
 - Introduce a new constant ADMIN_UPGRADE_TIMEOUT_INTERVAL
 - Replace magic numbers with ADMIN_UPGRADE_TIMEOUT_INTERVAL

santiph commented:
Approved
